### PR TITLE
separate rate limit for API

### PR DESCRIPTION
**Pull request by alon-e**
allows setting rate via API call: `setApiRateLimit`
fixes limit counter,
uses a larger rate limit window.

the limit is off by default.

### CODE REVIEW

File: src/main/java/com/iota/iri/service/API.java [hunk 1]
```diff
@@ -91,7 +91,7 @@
     private final int maxGetTrytes;
     private final int maxBodyLength;
     private final double newTransactionsRateLimit;
-    private final double newTransactionsLimit;
+    private double newTransactionsLimit;
```
Comment (th0br0) [inline src/main/java/com/iota/iri/service/API.java]: could we put this below the final vars

File: src/main/java/com/iota/iri/service/API.java [hunk 2]
```diff
@@ -354,7 +363,7 @@ private boolean isBelowNewTransactionLimit(InetAddress sourceAddress, int size)
             broadcastStoreTimer.set(now);
         }
 
-        if(broadcastStoreCounters.putIfAbsent(sourceAddress, new AtomicInteger(size)) != null) {
+        if(broadcastStoreCounters.putIfAbsent(sourceAddress, new AtomicInteger(size)) == null) { //not already in counter
             return size < newTransactionsLimit;
```
Comment (th0br0) [inline src/main/java/com/iota/iri/service/API.java]: this is a potential data race (parallel setApiRateLimit + read)
make this atomic?

File: src/main/java/com/iota/iri/service/API.java [hunk 3]
```diff
@@ -977,6 +998,11 @@ private boolean validTrytes(String trytes, int length, char zeroAllowed) {
         Matcher matcher = trytesPattern.matcher(trytes);
         return matcher.matches();
     }
+    
+    public void setApiRateLimit(double apiRateLimit) {
+        newTransactionsLimit = (apiRateLimit * Neighbor.newTransactionsWindow) / 1000;
```
Comment (th0br0) [inline src/main/java/com/iota/iri/service/API.java]: same race as above

File: src/main/java/com/iota/iri/service/API.java [hunk 4]
```diff
@@ -353,7 +354,8 @@ private AbstractResponse process(final String requestString, InetSocketAddress s
     }
 
     private boolean isBelowNewTransactionLimit(InetAddress sourceAddress, int size) {
-        if (newTransactionsLimit == 0) {
+        int newTransactionsLimitInt = newTransactionsLimit.get();
```
Comment (th0br0) [inline src/main/java/com/iota/iri/service/API.java]: nit: you might want to specify the order as relaxed here, that might give a minor speed boost (default is seq_cst)

**Comment by paulhandy** [comment]
rebased ( but it didn't mark it as merged )

**Pull request closed without merging**